Update Subscription creation API stub to actually create Subscriptions

wedamija opened the pull request:
This adds a `SubscriptionCreator` class that handles the domain logic of validating and creating a
subscription. Updates the endpoint to use this and return the id.

Adds a Partitioner to determine where we'll store a given subscription. At the moment this is
hardcoded, but in the future we can update it to fetch from the dataset.

I'll add a json schema for the API and do extra validation in a follow-up pr.

wedamija committed:
Update Subscription creation API stub to actually create Subscriptions

This adds a `SubscriptionCreator` class that handles the domain logic of validating and creating a
subscription. Updates the endpoint to use this and return the id.

Adds a Partitioner to determine where we'll store a given subscription. At the moment this is
hardcoded, but in the future we can update it to fetch from the dataset.

wedamija commented on snuba/subscriptions/data.py:
```diff
@@ -28,23 +27,28 @@ class Subscription:
         """
+        # XXX: We should move this somewhere better so that we don't need to import from
+        # web.views here.
+        from snuba.web.views import validate_request_content
```
Had to move this here due to circular imports, this might belong in `query/`

wedamija replied:
I'll look into doing this as a follow-up pr.

wedamija committed:
fix tests

wedamija commented on snuba/subscriptions/subscription.py:
```diff
@@ -0,0 +1,65 @@
+from datetime import datetime, timedelta
```
Not totally happy with the name of this file. These are kind of similar to the mediators that ecosystem uses, potentially we could call them something like that.

tkaemming replied:
I don't have a great idea either, I think this works for now though.

wedamija commented on snuba/subscriptions/subscription.py:
```diff
@@ -0,0 +1,65 @@
+        partition_id = DatasetSubscriptionPartitioner(self.dataset).build_partition_id(
+            subscription
+        )
+        RedisSubscriptionStore(redis_client, partition_id).create(subscription)
```
Hmm, does the key we pass to `RedisSubscriptionStore` need to include the dataset name in it? Thinking it probably does

tkaemming replied:
Yeah, it should. Probably just best to pass the `Dataset` to the constructor.

tkaemming commented on snuba/web/views.py:
```diff
@@ -364,12 +366,46 @@ def sdk_distribution(*, timer: Timer) -> Response:
+            {
+                "subscription_id": build_external_subscription_id(
+                    partition_id, subscription.id
+                )
```
I think that `(partition_id, subscription.id)` should be a standalone type (probably a namedtuple) and also the return value of `create`.

tkaemming commented on snuba/web/views.py:
```diff
@@ -364,12 +366,46 @@ def sdk_distribution(*, timer: Timer) -> Response:
-    key = uuid1().hex
+@util.time_request("subscription")
+def create_subscription(*, dataset: Dataset, timer: Timer):
+    data = parse_request_body(http_request)
```
Would be nice if this just used the `Subscription` codec.

wedamija commented on snuba/subscriptions/subscription.py:
```diff
@@ -0,0 +1,65 @@
+
+        # We want to test the query out here to make sure it's valid and can run
+        request = subscription.build_request(
+            self.dataset, datetime.now() - subscription.time_window, None, timer,
```
It's just arbitrary, but it runs the query for the period of time requested. No real need to do so, it doesn't make a difference to whether the query will run.

wedamija commented on snuba/subscriptions/subscription.py:
```diff
@@ -0,0 +1,65 @@
+        conditions: Sequence[Condition],
+        aggregations: Sequence[Aggregation],
+        time_window: timedelta,
+        resolution: timedelta,
```
That's a good justification for removing it... As long as we're able to still pass the subscription id to the results producer then we're fine, since I need that id to match updates to subscriptions on my end. We'll need to change the subscription store to return the subscription id and subscription when fetching all of them, which isn't a huge deal

wedamija committed:
Merge branch 'master' into danf/subscription_state

wedamija committed:
Move `time_window` and `resolution` validation to `Subscription`
Change `build_partition_id` to return int
Remove `id` from `Subscription`, and related changes
Pass `dataset` to `RedisSubscriptionStore` and include in key
Add `SubscriptionIdentifier` and return from `SubscriptionCreator.create`
Use `SubscriptionCodec` for `Subscription` validation in the api.
Make handling of `time_window` and `resolution` consistent (now always seconds when
serialized/deserialized)
Various mypy fixes

tkaemming commented on snuba/subscriptions/subscription.py:
```diff
@@ -27,39 +27,16 @@ def __init__(self, dataset: Dataset):
-        request = subscription.build_request(
-            self.dataset, datetime.now() - subscription.time_window, None, timer,
-        )
+        request = subscription.build_request(self.dataset, datetime.now(), None, timer,)
```
I think `datetime.now()` might need to be `datetime.utcnow()` for consistency here with code on the write path. Shouldn't matter in production, but local time might yield unexpected results in development. (Don't get me started about tzinfo.)

tkaemming commented on snuba/web/views.py:
```diff
@@ -379,32 +378,22 @@ def handle_subscription_error(exception: InvalidSubscriptionError):
+def build_external_subscription_id(identifier: SubscriptionIdentifier) -> str:
+    return "{}{}{}".format(
+        identifier.partition_id, SUBSCRIPTION_SEPARATOR, identifier.subscription_id
+    )
```
I would be OK with making this the `__str__` on `SubscriptionIdentifier` and just using that if you'd like.

wedamija replied:
I was thinking about that... It felt weird for `SubscriptionIdentifier` to be aware of the format that the api expects the identifier to be in.

tkaemming replied:
I think that's reasonable too, I don't have a strong opinion about it and it's relatively easy to change. TBH, might just be worth inlining it for now as an f-fstring, unless you plan in reusing it — I think it'd be equally clear that way. Fine as-is too if you'd rather.

tkaemming commented on snuba/subscriptions/data.py:
```diff
@@ -9,42 +9,59 @@
     time_window: timedelta
     resolution: timedelta
 
+    def __post_init__(self):
```
```suggestion
    def __post_init__(self) -> None:
```

wedamija replied:
Mypy never catches this, do I have something configured wrong?

tkaemming replied:
Hmm, do you not get this when running on c3b7c208cd6bc36b662b851b3edd0e3cfadf2767?

```
ted@veneno % python -m mypy --strict snuba/subscriptions | grep snuba/subscriptions
snuba/subscriptions/data.py:32: error: Function is missing a return type annotation
snuba/subscriptions/data.py:32: note: Use "-> None" if function does not return a value
```

wedamija replied:
Yeah, I didn't have `--strict`, should be able to catch these going forward, thanks!

tkaemming commented on snuba/web/views.py:
```diff
@@ -363,12 +365,36 @@ def sdk_distribution(*, timer: Timer) -> Response:
+def create_subscription(*, dataset: Dataset, timer: Timer):
+    subscription = SubscriptionCodec().decode(http_request.data)
+    # TODO: Check for valid queries with fields that are invalid for subscriptions. For
+    # example date fields and aggregates.
```
Should this happen in `Subscription.__post_init__`?

wedamija replied:
Probably, yeah.

tkaemming commented on snuba/subscriptions/partitioner.py:
```diff
@@ -0,0 +1,28 @@
+    PARTITION_COUNT = 64
+
+    def __init__(self, dataset: Dataset):
+        self.dataset = dataset
```
I'd consider making this private unless you need to expose it.

tkaemming commented on snuba/subscriptions/store.py:
```diff
@@ -39,39 +39,38 @@ class RedisSubscriptionStore:
         self.client = client
+        self.dataset = dataset
         self.key = key
         self.codec = SubscriptionCodec()
```
It'd probably be good to loop back and make these private later, but better for them to be consistent right now.

tkaemming reviewed (approved):
This looks really good. There are a couple minor things inline you might want to address before merging but nothing too serious.

wedamija committed:
Update snuba/subscriptions/data.py

Co-Authored-By: ted kaemming <ted@getsentry.com>

wedamija committed:
Merge branch 'master' into danf/subscription_state

wedamija committed:
Code review feedback
